staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -0,0 +1,370 @@
+		if requestInfo.APIGroup == "" {
+			gvr.Group = "core"
+		}
+
+		// find servers that are capable of serving this request
+		serviceableByResp, err := h.findServiceableByServers(gvr, serverId, reconciler)
+		if err != nil {
+			// this means that resource is an aggregated API or a CR since it wasn't found in SV informer cache, pass as it is
+			klog.Infof(fmt.Sprintf("no StorageVersion/APIServerID found for the GVR: %v skipping proxying", gvr))
+			handler.ServeHTTP(w, r)
+			return
+		}
+		// found the gvr locally, pass request to the next handler in local apiserver
+		if serviceableByResp.locallyServiceable {
+			klog.V(4).Infof("resource can be served locally, skipping proxying")
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		gv := schema.GroupVersion{Group: gvr.Group, Version: gvr.Version}
+
+		// if no apiservers were found that could serve the request, serve 404
+		if len(serviceableByResp.peerIPs) == 0 {
+			klog.Errorf(fmt.Sprintf("GVR %v is not served by anything in this cluster", gvr))
+			responsewriters.ErrorNegotiated(apierrors.NewNotFound(schema.GroupResource{Group: gvr.Group, Resource: gvr.Resource},
+				fmt.Sprintf("%s.%s", gvr.Group, gvr.Resource)), s, gv, w, r)
+			return
+		}
+
+		// otherwise, randomly select an apiserver
+		rand := rand.Intn(len(serviceableByResp.peerIPs))
+		destServerIP := serviceableByResp.peerIPs[rand]
+
+		if destServerIP == "" {
+			klog.Errorf("failed to serve request: Found no endpoints in server leases for the remote apiserver")
+			responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Found no endpoints in servers leases for the remote apiserver"), s, gv, w, r)
+			return
+		}
+
+		// check ip format
+		_, _, err = net.SplitHostPort(destServerIP)
+		if err != nil {
+			klog.ErrorS(err, "error getting ip and port info of the remote server while proxying")
+			responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Error getting ip and port info of the remote server while proxying"), s, gv, w, r)
+			return
+		}
+
+		h.proxyRequestToDestinationAPIServer(r, w, destServerIP)
+
+	})
+}
+
+func (h *peerProxyHandler) findServiceableByServers(gvr schema.GroupVersionResource, localAPIServerId string, reconciler reconcilers.PeerEndpointLeaseReconciler) (serviceableByResponse, error) {
+
+	apiserversi, ok := h.svMap.Load(gvr)
+
+	// no value found for the requested gvr in svMap
+	if !ok || apiserversi == nil {
+		return serviceableByResponse{}, fmt.Errorf("no StorageVersions found for the GVR: %v", gvr)
+	}
+	apiservers := apiserversi.(*sync.Map)
+	response := serviceableByResponse{}
+	var peerServerIPs []string
+	apiservers.Range(func(key, value interface{}) bool {
+		apiserverKey := key.(string)
+		if apiserverKey == localAPIServerId {
+			response.locallyServiceable = true
+			// stop iteration
+			return false
+		}
+
+		ip, err := reconciler.GetEndpoint(apiserverKey)
```

Comment:
we have `ip` as variable is really "ip:port", can we use `ipPort` or `hostPort` on the variables to reflect better the content?

Reply:
done.